plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java

```diff
@@ -0,0 +1,321 @@
+            return Optional.of(decimalColumnMapping(createDecimalType(precision, scale),
+                    getDecimalRoundingMode(session)));
+        }
+        if (jdbcTypeName.equals("String")) {
+            return Optional.of(varcharColumnMapping(createUnboundedVarcharType()));
+        }
+        if (jdbcTypeName.equals("Date")) {
+            return Optional.of(dateColumnMapping());
+        }
+        if (jdbcTypeName.equals("DateTime")) {
+            // clickhouse not implemented for type=class java.time.LocalDateTime
+            // TODO replace it using timestamp relative function after clickhouse adds support for LocalDateTime
+            return Optional.of(varcharColumnMapping(createUnboundedVarcharType()));
+        }
+        if (getUnsupportedTypeHandling(session) == CONVERT_TO_VARCHAR) {
+            return mapToUnboundedVarchar(typeHandle);
+        }
+        // unknown type ?
+        return Optional.empty();
+    }
+
+    @Override
+    public WriteMapping toWriteMapping(ConnectorSession session, Type type)
+    {
+        if (type == BOOLEAN) {
+            return WriteMapping.booleanMapping("UInt8", booleanWriteFunction());
+        }
+        if (type == TINYINT) {
+            return WriteMapping.longMapping("Int8", tinyintWriteFunction());
+        }
+        if (type == SMALLINT) {
+            return WriteMapping.longMapping("Int16", smallintWriteFunction());
+        }
+        if (type == INTEGER) {
+            return WriteMapping.longMapping("Int32", integerWriteFunction());
+        }
+        if (type == BIGINT) {
+            return WriteMapping.longMapping("Int64", bigintWriteFunction());
+        }
+        if (type == REAL) {
+            return WriteMapping.longMapping("Float32", realWriteFunction());
+        }
+        if (type == DOUBLE) {
+            return WriteMapping.doubleMapping("Float64", doubleWriteFunction());
+        }
+        if (type instanceof DecimalType) {
+            DecimalType decimalType = (DecimalType) type;
+            String dataType = format("decimal(%s, %s)", decimalType.getPrecision(), decimalType.getScale());
+            if (decimalType.isShort()) {
+                return WriteMapping.longMapping(dataType, shortDecimalWriteFunction(decimalType));
+            }
+            return WriteMapping.sliceMapping(dataType, longDecimalWriteFunction(decimalType));
+        }
+        if (type instanceof CharType || type instanceof VarcharType) {
+            return WriteMapping.sliceMapping("String", varcharWriteFunction());
+        }
+        if (type instanceof VarbinaryType) {
+            return WriteMapping.sliceMapping("String", varbinaryWriteFunction());
+        }
+        if (type == DATE) {
+            return WriteMapping.longMapping("date", dateWriteFunction());
+        }
+        if (type instanceof TimestampType) {
+            return WriteMapping.longMapping("DateTime", timestampWriteFunction(TIMESTAMP_MILLIS));
```

Comment:
What about other precisions?

I'd suggest to remove timestamps from this PR and do them in a follow-up since they require multiple passes to implement and review correctly. It will help merge this PR sooner.